components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1467,6 +1470,50 @@ where
         }
     }
 
+    // If lease expired, we will send a noop read index to renew lease.
+    fn try_renew_leader_lease(&mut self) {
+        if !self.fsm.peer.is_leader() {
+            return;
+        }
+        if let Err(e) = self.fsm.peer.pre_read_index() {
+            debug!(
+                "prevent unsafe read index to renew leader lease";
+                "region_id" => self.region_id(),
+                "peer_id" => self.fsm.peer_id(),
+                "err" => ?e,
+            );
+            return;
+        }
+
+        let current_time = *self.ctx.current_time.get_or_insert_with(monotonic_raw_now);
+        let max_lease = self.ctx.cfg.raft_store_max_leader_lease();
+        let renew_bound = current_time + self.ctx.cfg.check_leader_lease_interval();
+        // We need to propose a read index request if current lease can't cover till next tick
+        let need_propose = match self.fsm.peer.leader_lease.inspect(Some(renew_bound)) {
+            LeaseState::Expired => {
+                self.fsm.peer.pending_reads.back().map_or(true, |read| {
+                    // If there is any read index whose lease can cover till next heartbeat
+                    // then we don't need to propose a new one
+                    read.propose_time + max_lease < renew_bound
+                }) && self.fsm.peer.proposals.back().map_or(true, |proposal| {
+                    // If there is any write whose lease can cover till next heartbeat
+                    // then we don't need to propose a new one
+                    proposal
+                        .propose_time
+                        .map_or(true, |propose_time| propose_time + max_lease < renew_bound)
+                })
+            }
+            _ => false,
+        };
+        if need_propose {
+            let (id, dropped) = self.fsm.peer.propose_read_index(None, None);
+            if !dropped {
```

Comment:
There should be metrics to monitor failed attempts.

Reply:
Added a counter in raft metrics